### ISSUE TITLE
m.route redirects to correct route, but renders the wrong view

### ISSUE THREAD

**Issue by vyrak** [lhorie/mithril.js#687]
I have a case where I'd like to redirect to another route if a certain condition is met when a component's controller is being initialized. Below, is the code that demonstrates the case.

```javascript
var root = {
  controller: function() {
    var onclick = function() { m.route('/1'); };

    return {onclick: onclick};
  },
  view: function(c) {
    return m('h1', {onclick: c.onclick}, 'Root');
  }
};

var page = {
  controller: function() {
    var number, next, onclick;
    number = parseInt(m.route.param('number'), 10);

    // Some arbitrary condition to redirect to another route
    if (number > 2) { m.route('/'); }

    next = number + 1;
    onclick = function() { m.route('/' + next); };

    return {
      number: number,
      onclick: onclick
    };
  },
  view: function(c) {
    return m('h1', {onclick: c.onclick}, 'Page ' + c.number);
  }
};

m.route.mode = 'hash';
m.route(document.getElementById('content'), '/', {
  '/': root,
  //'/:number': page            // redirects to '/' and renders root.view as expected
  '/:number': {
    view: function() {
      return m.component(page); // redirects to '/' but renders this view instead
    }
  }
});

```

When the `page` component gets mapped directly to the `/:number` route, the application has the correct route in the url & renders the `root` component correctly as expected. However, when I wrap the `page` component with an outer component that tries to render it using `m.component`, the application routes to the correct url, but does not render the `root` component.

I'm using the version `0.2.0` and testing on Chrome 43.0.2357.65.

Edit:

Here's a codepen... http://codepen.io/anon/pen/rVYdeQ?editors=101

**Comment by pelonpelon** [comment]
It's a little complicated but if you follow the console log in this jsbin you might see what's happening.
http://jsbin.com/yoniqi/1/edit?js,console,output

Mithril puts calls to redraw in a queue. Every 1/16th of a second it fires all accumulated redraws in one. When you click on Page:/2 there's a redraw called for the click (page) and a redraw called for m.route('/'). Although the root controller gets called (and you get the correct url in the browser), the page view is called and the redraw completes before the root view gets called.

There is a place in my code where you can uncomment a setTimeout. This causes the root to redirect in a later 1/16th of a second frame. It gets a queue all its own and you get the results you expect (with a slight delay).

I suggest you not call route() directly from a controller. In your case you should move it to the onclick handler. Do your `if (number > 2)` logic there.

BTW: your pen and the code above don't match exactly

**Comment by vyrak** [comment]
Thanks for the extra pair of eyes. I did realize I could use `setTimeout` to delay the redirection, but didn't know how it worked in relation to Mithril's until your explanation. I ended up tweaking the timeout callback so the perceived delay is not there.

```javascript
m.startComputation();
setTimeout(function(){
  m.route('/');
  m.endComputation();
});
```
Maybe it's worth having this logic in Mithril's `m.route(path)` method.

The reason I use `m.route` in the controller is because when the component is initialized, it looks to a service to resolve the component's data; if the data is not available, it'd redirect.

If that logic was in the `onclick` handler, the data from the service would have to be stashed somewhere to be retrieved later by the component when it is initialized after a redirect. Even then, the object doing the stashing would have to "guarantee" that that data was available (based on the route's params) to render said component; otherwise, it brings us back to the same scenario. Or maybe I'm thinking about it all wrong :p

In regards to the code mismatch between here and the pen, I only tweaked the templates to render the route (I originally pasted it from my local machine, where Chrome took care of displaying that). Damage done. Oh well.
